File: src/main/java/io/github/starwishsama/namelessbot/objects/user/BotUser.java
package io.github.starwishsama.namelessbot.objects.user;

import cc.moecraft.icq.user.User;
import io.github.starwishsama.namelessbot.BotConstants;
import io.github.starwishsama.namelessbot.BotMain;
import io.github.starwishsama.namelessbot.enums.UserLevel;
import io.github.starwishsama.namelessbot.utils.BotUtils;
import lombok.Data;

import java.util.Calendar;
import java.util.LinkedList;
import java.util.List;
import java.util.UUID;

@Data
public class BotUser {
    private long userQQ;
    private UUID userUUID;
    private Calendar lastCheckInTime = Calendar.getInstance();
    private double checkInPoint;
    private int checkInTime = 0;
    private String bindServerAccount;
    private int msgVL;
    private String r6sAccount;
    private UserLevel level = UserLevel.USER;
    private int randomTime = 50;
    private Pet pet;
    private long checkInGroup;
    private List<Long> biliSubs = new LinkedList<>();

    public BotUser(long qq){
        userQQ = qq;
        userUUID = BotUtils.generateUuid();
    }

    public static boolean isUserExist(long qq) {
        return getUser(qq) != null;
    }

    public static boolean isBotAdmin(long id){
        return BotUtils.getLevel(id).ordinal() > 1;
    }

    public static boolean isBotOwner(long id){
        return BotUtils.getLevel(id) == UserLevel.OWNER || BotConstants.cfg.getOwnerID() == id;
    }

    public static boolean isBotOwner(User sender){
        return isBotOwner(sender.getId());
    }

    public static boolean isBotAdmin(User sender){
        return isBotAdmin(sender.getId());
    }

    public void decreaseTime(){
        if (level == UserLevel.USER) {
            this.randomTime--;
        }
    }

    public void decreaseTime(int time){
        if (level == UserLevel.USER) {
            for (int i = 0; i < time; i++) {
                this.randomTime--;
            }
        }
    }

    public void updateTime() {
        if (level == UserLevel.USER && randomTime < 20) {
            this.randomTime++;
        }
    }

    public void addPoint(double point) {
        checkInPoint = checkInPoint + point;
    }

    public void addTime(int time) {
        if (level == UserLevel.USER && randomTime < 20) {
            randomTime = randomTime + time;
        }
    }

    public void cost(double point){
        checkInPoint = checkInPoint - point;
    }

    public static BotUser quickRegister(long id){
        BotUser user = new BotUser(id);
        BotConstants.users.add(user);
        return user;
    }

    public static BotUser getUser(Long qq){
        if (BotConstants.users != null){
            for (BotUser user : BotConstants.users){
                if (user.getUserQQ() == qq) {
                    return user;
                }
            }
        } else {
            BotMain.getLogger().warning("在获取 QQ 号为 " + qq + " 的用户数据时出现了问题: 用户列表为空");
        }
        return null;
    }

    public static BotUser getUser(User sender){
        if (BotConstants.users != null){
            for (BotUser user : BotConstants.users){
                if (user.getUserQQ() == sender.getId()) {
                    return user;
                }
            }
        } else {
            BotMain.getLogger().warning("在获取 QQ 号为 " + sender.getId() + " 的用户数据时出现了问题: 用户列表为空");
        }
        return null;
    }
}

File: src/main/kotlin/io/github/starwishsama/comet/objects/BotUser.kt
package io.github.starwishsama.comet.objects

import com.fasterxml.jackson.annotation.JsonProperty
import io.github.starwishsama.comet.BotVariables
import io.github.starwishsama.comet.enums.UserLevel
import io.github.starwishsama.comet.objects.wrapper.MessageWrapper
import java.time.LocalDateTime

data class BotUser(
    @JsonProperty("userQQ")
    val id: Long,
    var lastCheckInTime: LocalDateTime = LocalDateTime.MIN,
    var checkInPoint: Double = 0.0,
    var checkInTime: Int = 0,
    var bindServerAccount: String = "",
    var r6sAccount: String = "",
    var level: UserLevel = UserLevel.USER,
    var commandTime: Int = 100,
    var checkInGroup: Long = 0,
    var lastExecuteTime: Long = -1,
    private val permissions: MutableList<String> = mutableListOf(),
    val savedContents: MutableList<MessageWrapper> = mutableListOf()
) {
    fun decreaseTime() {
        if (level <= UserLevel.VIP) {
            commandTime--
        }
    }

    fun decreaseTime(time: Int) {
        if (level <= UserLevel.VIP) {
            commandTime -= time
        }
    }

    fun addPoint(point: Double) {
        checkInPoint += point
    }

    fun addTime(time: Int, bypass: Boolean = false) {
        if (level == UserLevel.USER && (commandTime <= 1000 || bypass)) {
            commandTime += time
        }
    }

    fun costPoint(point: Double) {
        checkInPoint -= point
    }

    fun plusDay() {
        checkInTime++
    }

    fun resetDay(){
        checkInTime = 1
    }

    fun hasPermission(permission: String): Boolean {
        return permissions.contains(permission) || isBotOwner()
    }

    fun getPermissions(): String = buildString {
        this@BotUser.permissions.forEach {
            append("$it ")
        }
    }.trim()

    /**
     * 比较权限组
     * @return 自己的权限组是否大于等于需要比较的权限组
     */
    fun compareLevel(cmdLevel: UserLevel): Boolean {
        return this.level >= cmdLevel
    }

    fun isBotAdmin(): Boolean {
        return level >= UserLevel.ADMIN || isBotOwner()
    }

    fun isBotOwner(): Boolean {
        return level == UserLevel.OWNER
    }

    fun addPermission(permission: String) {
        permissions.plusAssign(permission)
    }

    /**
     * 判断是否签到过了
     *
     * @author NamelessSAMA
     * @return 是否签到
     */
    fun isChecked(): Boolean {
        val now = LocalDateTime.now()
        val period = lastCheckInTime.toLocalDate().until(now.toLocalDate())

        return period.days == 0
    }

    companion object {
        fun isBotOwner(id: Long): Boolean {
            return getUser(id)?.level == UserLevel.OWNER
        }

        fun quickRegister(id: Long): BotUser {
            val user = BotUser(id)
            return BotVariables.users.putIfAbsent(id, user) ?: user
        }

        fun getUser(id: Long): BotUser? {
            return BotVariables.users[id]
        }

        fun getUserOrRegister(qq: Long): BotUser = getUser(qq) ?: quickRegister(qq)
    }
}